File: Structural/Lab12/Typescript/src/Entities/real_estate_flyweight_factory.ts
import { RealEstateFlyweight } from "./real_state_flyweight";

export class RealEstateFlyweightFactory {
    private flyweights: { [key: string]: RealEstateFlyweight } = {};

    constructor(...args: RealEstateFlyweight[]) {
        args.forEach(elem => {
            this.flyweights[this.getKey(elem)] = elem;
        });
    }

    private getKey(key: RealEstateFlyweight): string {
        const elements = [
            key.furnished.toString(),
            key.parking.toString(),
            key.balcony.toString(),
            key.elevator.toString(),
            key.airConditioning.toString(),
            key.isUtilitiesIncluded.toString()
        ];

        return elements.join('_');
    }

    public getFlyweight(sharedState: RealEstateFlyweight): RealEstateFlyweight {
        const key = this.getKey(sharedState);

        if (!this.flyweights.hasOwnProperty(key)) {
            console.log('RealEstateFlyweightFactory: Can\'t find a flyweight, creating new one.');
            this.flyweights[key] = sharedState;
        } else {
            console.log('FRealEstateFlyweightFactory: Reusing existing flyweight.');
        }
        return this.flyweights[key];
    }


}

File: Structural/Lab12/C#/Entities/RealEstateFlyweightFactory.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Text;

namespace C_.Entities
{
    public class RealEstateFlyweightFactory
    {
        private Dictionary<string, RealEstateFlyweight> flyweights = new Dictionary<string, RealEstateFlyweight>();

        public RealEstateFlyweightFactory(params RealEstateFlyweight[] args)
        {
            foreach (var elem in args)
            {
                flyweights.Add(GetKey(elem), elem);
            }
        }

        public string GetKey(RealEstateFlyweight key)
        {
            List<string> elements = new List<string>
            {
                key.Furnished.ToString(),
                key.Parking.ToString(),
                key.Balcony.ToString(),
                key.Elevator.ToString(),
                key.AirConditioning.ToString(),
                key.IsUtilitiesIncluded.ToString()
            };


            return string.Join("_", elements);
        }


        public RealEstateFlyweight GetFlyweight(RealEstateFlyweight sharedState)
        {
            string key = this.GetKey(sharedState);

            if (flyweights.Where(t => t.Key == key).Count() == 0)
            {
                Console.WriteLine("RealEstateFlyweightFactory: Can't find a flyweight, creating new one.");
                this.flyweights.Add(key, sharedState);
            }
            else
            {
                Console.WriteLine("FRealEstateFlyweightFactory: Reusing existing flyweight.");
            }
            return this.flyweights.Where(t => t.Key == key).FirstOrDefault().Value;
        }


    }
}
